File: tests/e-rules/comprehensive_enhanced_test.js
// Comprehensive test file for enhanced pattern features in JavaScript

// Test dangerous eval comprehensive
function processUserInput(userCode) {
    // Should match - eval with string concatenation outside sandbox/try-catch
    eval("var result = " + userCode + ";");
    
    // Should match - another concatenation pattern
    return eval("process_" + operation + "(" + data + ")");
}

function sandboxedExecution(code) {
    // Should NOT match - inside sandbox function
    function sandboxHelper() {
        eval(code);
    }
    sandboxHelper();
}

function sandboxProcessor(input) {
    // Should NOT match - inside sandbox function
    eval("processInput(" + input + ")");
}

function safeEvaluation(code) {
    try {
        // Should NOT match - inside try-catch block
        eval("var temp = " + code + ";");
    } catch (error) {
        console.error("Eval failed:", error);
    }
}

function testEvalFunction() {
    // Should NOT match - test function excluded by pattern-not-regex
    eval("var testVar = " + testData + ";");
}

function mockEvaluation() {
    // Should NOT match - mock function excluded by pattern-not-regex
    eval("var mockResult = " + mockData + ";");
}

// Complex nested scenarios
function complexProcessing() {
    if (isProduction) {
        // Should match - eval with concatenation outside safety blocks
        eval("config." + property + " = " + value + ";");
        
        try {
            // Should NOT match - inside try-catch
            eval("backup." + key + " = " + backup_value + ";");
        } catch (e) {
            handleError(e);
        }
    }
}

// Edge cases
function dynamicExecution() {
    const code = "function_" + type + "(" + params + ")";
    
    // Should match - eval with concatenated string
    eval(code);
}

function conditionalEval() {
    if (allowDynamicCode) {
        // Should match - eval with concatenation
        eval("handler_" + eventType + "(event)");
    }
}

// Safe alternatives (should not match)
function safeAlternatives() {
    // These should not match as they don't use string concatenation
    eval("staticCode()");
    eval(precomputedCode);
    
    // Function calls that look similar but aren't eval
    evaluate("some_" + "code");
    execute("command_" + "args");
}

File: tests/e-rules/comprehensive_enhanced_test.py
# Comprehensive test file for enhanced pattern features

import requests
import sqlite3
import os

# Test complex security check
def vulnerable_function():
    # Should match - execute with string concatenation outside try-catch
    db.execute("SELECT * FROM users WHERE id = " + user_id)
    
    # Should match - eval with concatenation outside try-catch
    result = eval("process_" + operation + "(data)")

def test_function():
    # Should NOT match - excluded by pattern-not-regex (test_.*)
    db.execute("SELECT * FROM test_table WHERE id = " + test_id)

def safe_function():
    try:
        # Should NOT match - inside try-catch block
        db.execute("DELETE FROM users WHERE name = " + username)
    except Exception as e:
        handle_error(e)

# Test unsafe HTTP outside dev
def production_code():
    # Should match - HTTP request outside development context
    response = requests.get("http://api.production.com/data", headers=headers)

def development_code():
    if os.environ.get("ENV") == "development":
        # Should NOT match - inside development context
        response = requests.get("http://api.staging.com/data", headers=headers)

def test_api_call():
    # Should NOT match - test function excluded
    response = requests.get("http://api.test.com/data", headers=headers)

def local_testing():
    # Should NOT match - localhost excluded by pattern-not-regex
    response = requests.get("http://localhost:8080/api", headers=headers)
    
    # Should NOT match - 127.0.0.1 excluded by pattern-not-regex
    response = requests.get("http://127.0.0.1:3000/data", headers=headers)

# Test advanced SQL injection
def database_operations():
    cursor = sqlite3.cursor()
    
    # Should match - string concatenation outside transaction
    query = "SELECT * FROM users WHERE name = '" + username + "'"
    cursor.execute(query, params)
    
    # Should match - another concatenation pattern
    cursor.execute("DELETE FROM " + table + " WHERE id = '" + user_id + "'", params)

def safe_database_operations():
    connection = sqlite3.connect("db.sqlite")
    with connection.begin():
        cursor = connection.cursor()
        # Should NOT match - inside transaction block
        query = "SELECT * FROM users WHERE name = '" + username + "'"
        cursor.execute(query, params)

def test_database():
    # Should NOT match - test function excluded
    cursor = sqlite3.cursor()
    query = "SELECT * FROM test_users WHERE name = '" + test_name + "'"
    cursor.execute(query, params)
